### ISSUE TITLE
[Bug] [NPU][MoE] AscendTPDispatcher downcasts FP32 routing weights to the hidden-state dtype

### ISSUE THREAD

**Issue by Guozhong23** [sgl-project/sglang#39351]
### Checklist

- [ ] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [ ] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [ ] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [ ] Please use English. Otherwise, it will be closed.

### Describe the bug

The NPU TopK backend returns `topk_weights` in FP32, but `AscendTPDispatcher.dispatch()` immediately casts them to the hidden-state dtype:

```
topk_weights, topk_ids, _ = topk_output
topk_weights = topk_weights.to(hidden_states.dtype)
```

For BF16 inference this converts routing probabilities to BF16 before `npu_moe_finalize_routing`. This defeats the FP32 contract established by the TopK layer and introduces avoidable numerical error in every MoE layer.
This is especially visible for sigmoid routing without renormalization, small routing weights, or models where several selected experts have close scores. Repeating the quantization error through many MoE layers can change logits and generated tokens.
The current behavior is also inconsistent across NPU dispatchers: the generic/NPU TopK path explicitly converts its output to FP32, while the Ascend TP dispatcher discards that precision.

### Reproduction

The issue is visible directly in:
python/sglang/srt/layers/moe/token_dispatcher/ascend_tp.py
AscendTPDispatcher.dispatch()

A minimal regression test can construct BF16 hidden states and an FP32 `TopKOutput`, run the dispatcher, and check the dtype stored in `AscendTPDispatchOutput`:

```
assert topk_output.topk_weights.dtype == torch.float32
dispatch_output = dispatcher.dispatch(hidden_states_bf16, topk_output)
assert dispatch_output.topk_weights.dtype == torch.float32
```

Current result:
`dispatch_output.topk_weights.dtype == torch.bfloat16`

For an end-to-end reproduction, run an MoE model with BF16 hidden states and sigmoid/correction-bias routing on the Ascend TP dispatcher, and compare the finalizer input or layer output against the torch-native FP32 routing-weight reference.

### Environment

[root@TENCENT64-124 hw_sgz]# pip list
Package                                  Version                             Editable project location
---------------------------------------- ----------------------------------- ---------------------------------------------------------------------------------
absl-py                                  2.5.0
accelerate                               1.14.0
addict                                   2.4.0
affinity-sched                           26.1.0
aiofiles                                 25.1.0
aiohappyeyeballs                         2.7.1
aiohttp                                  3.14.1
aiosignal                                1.4.0
airportsdata                             20260803
ais-bench-benchmark                      3.1.20260630
aliyun-python-sdk-core                   2.16.0
aliyun-python-sdk-kms                    2.16.5
annotated-doc                            0.0.4
annotated-types                          0.7.0
anthropic                                0.117.0
antlr4-python3-runtime                   4.13.2
anyio                                    4.14.2
apache-tvm-ffi                           0.1.12
arctic_inference                         0.1.1
asc_op_compile_base                      0.1.0
asc_opc_tool                             0.1.0
astor                                    0.8.1
asttokens                                3.0.2
attrs                                    24.2.0
auto_tune                                0.1.0
av                                       18.0.0
blake3                                   1.0.9
blinker                                  1.9.0
blobfile                                 3.0.0
bracex                                   3.0.1
build                                    1.5.1
cache_dit                                1.3.5
cachetools                               7.1.4
cann_ops_transformer                     1.0.0
cannbotdsl                               0.3.dev527+gb9fc3a227.d20260908
cannsim                                  0.1.0
cbor2                                    6.1.3
certifi                                  2026.6.17
cffi                                     2.1.0
charset-normalizer                       3.4.9
click                                    8.4.2
cloudpickle                              3.1.2
cmake                                    4.4.0
colorama                                 0.4.6
colorlog                                 6.12.0
compressed-tensors                       0.17.0
contourpy                                1.3.3
cpm-kernels                              1.0.11
crcmod                                   1.7
cryptography                             49.0.0
cycler                                   0.12.1
dataflow                                 0.0.1
datasets                                 4.8.4
decorator                                5.1.1
deep_ep                                  1.0.0+1cda4d1a.cann.9.1.0.b230
defusedxml                               0.7.1
depyf                                    0.20.0
detect-installer                         0.1.0
diffusers                                0.37.0
dill                                     0.3.8
diskcache                                5.6.3
distro                                   1.9.0
dnspython                                2.8.0
docstring_parser                         0.18.0
dotenv                                   0.9.9
easydict                                 1.13
editdistance                             0.8.1
einops                                   0.8.2
email-validator                          2.3.0
es_cv                                    1.0.0
es_hccl                                  1.0.0
es_math                                  1.0.0
es_nn                                    1.0.0
es_transformer                           1.0.0
evalscope                                1.6.1
evaluate                                 0.4.6
execnet                                  2.1.2
executing                                2.2.1
fastapi                                  0.123.10
fastapi-cli                              0.0.32
fastapi-cloud-cli                        0.22.2
fastar                                   0.11.0
filelock                                 3.31.0
fire                                     0.7.1
Flask                                    3.1.3
fonttools                                4.63.0
frozenlist                               1.8.0
fsspec                                   2025.3.0
func-timeout                             4.3.5
fuzzywuzzy                               0.18.0
ge-py                                    0.0.1
ge-py-pass-bridge                        0.0.1
gevent                                   26.5.0
gguf                                     0.19.0
googleapis-common-protos                 1.75.0
gradio_client                            2.5.0
greenlet                                 3.5.3
grpcio                                   1.83.0
grpcio-health-checking                   1.83.0
grpcio-reflection                        1.83.0
h11                                      0.16.0
h2                                       4.3.0
h5py                                     3.16.0
hccl                                     0.1.0
hf_transfer                              0.1.9
hf-xet                                   1.5.2
hpack                                    4.2.0
httpcore                                 1.0.9
httptools                                0.8.0
httpx                                    0.28.1
httpx-sse                                0.4.3
huggingface_hub                          1.24.0
human-eval                               1.0.3
Hypercorn                                0.18.0
hyperframe                               6.1.0
idna                                     3.18
ijson                                    3.5.1
imageio                                  2.36.0
imageio-ffmpeg                           0.5.1
immutabledict                            4.3.1
importlib_metadata                       9.0.0
iniconfig                                2.3.0
interegular                              0.3.3
ipython                                  9.16.1
ipython_pygments_lexers                  1.1.1
itsdangerous                             2.2.0
janus                                    2.0.0
jedi                                     0.20.0
jieba                                    0.42.1
Jinja2                                   3.1.6
jiter                                    0.16.0
jmespath                                 0.10.0
joblib                                   1.5.3
json5                                    0.15.0
jsonlines                                4.0.0
jsonschema                               4.26.0
jsonschema-specifications                2025.9.1
kiwisolver                               1.5.0
langdetect                               1.0.9
lark                                     1.2.2
latex2sympy2_extended                    1.0.6
lazy-loader                              0.5
Levenshtein                              0.27.3
llguidance                               1.7.6
llm_datadist                             0.0.1
llm_datadist_v1                          0.0.1
llvmlite                                 0.48.0
lm-format-enforcer                       0.11.3
loguru                                   0.7.3
lxml                                     6.1.1
Markdown                                 3.10.3
markdown-it-py                           4.2.0
MarkupSafe                               3.0.3
math-verify                              0.5.2
matplotlib                               3.11.1
matplotlib-inline                        0.2.2
mcp                                      1.28.1
mdurl                                    0.1.2
memcache_hybrid                          1.2.0
memfabric_hybrid                         1.2.0
mindstudio-kl                            26.1.0
mindstudio-kpp                           26.1.0
mindstudio-opgen                         26.1.0
mindstudio-opst                          26.0.0
mistral_common                           1.11.6
mmengine-lite                            0.10.7
model-hosting-container-standards        0.1.16
modelscope                               1.39.0
modelscope-hub                           0.2.0
mooncake-transfer-engine-npu             0.3.11.post1
more-itertools                           11.1.0
moviepy                                  2.2.1
mpmath                                   1.3.0
ms_service_profiler                      26.0.0
msgpack                                  1.2.1
msgspec                                  0.21.1
msmodelslim                              26.1.0
msobjdump                                0.1.0
mspti                                    26.1.0
multidict                                6.7.1
multiprocess                             0.70.16
narwhals                                 2.24.0
nest-asyncio                             1.6.0
networkx                                 3.6.1
ninja                                    1.13.0
nltk                                     3.10.0
numba                                    0.66.0
numpy                                    1.26.4
op_compile_tool                          0.1.0
opc_tool                                 0.1.0
openai                                   2.6.1
openai-benchmark                         1.0.0                               /mnt/share/release_openai_benchmark/release_openai_benchmark
openai-harmony                           0.0.4
opencv-python-headless                   5.0.0.93
opentelemetry-api                        1.44.0
opentelemetry-exporter-otlp              1.44.0
opentelemetry-exporter-otlp-proto-common 1.44.0
opentelemetry-exporter-otlp-proto-grpc   1.44.0
opentelemetry-exporter-otlp-proto-http   1.44.0
opentelemetry-proto                      1.44.0
opentelemetry-sdk                        1.44.0
opentelemetry-semantic-conventions       0.65b0
opentelemetry-semantic-conventions-ai    0.5.1
optype_collector                         0.1.0
orjson                                   3.11.9
oss2                                     2.19.1
outlines                                 0.1.11
outlines_core                            0.1.26
overrides                                7.7.0
packaging                                26.2
pandas                                   3.0.4
pandas-stubs                             3.0.3.260530
parso                                    0.8.7
partial-json-parser                      0.2.1.1.post7
pexpect                                  4.9.0
pillow                                   11.2.1
pip                                      26.2.1
platformdirs                             4.10.1
plotly                                   6.9.0
pluggy                                   1.6.0
portalocker                              3.2.0
prettytable                              3.18.0
priority                                 2.0.0
proglog                                  0.1.12
prometheus_client                        0.25.0
prometheus-fastapi-instrumentator        8.0.2
prompt_toolkit                           3.0.53
propcache                                0.5.2
protobuf                                 7.35.1
psutil                                   6.0.0
ptyprocess                               0.7.0
pure_eval                                0.2.3
py-cpuinfo                               9.0.0
py-spy                                   0.4.2
pyarrow                                  25.0.0
pybase64                                 1.4.3
pybind11                                 3.0.4
pycountry                                26.2.16
pycparser                                3.0
pycryptodome                             3.23.0
pycryptodomex                            3.23.0
pydantic                                 2.13.4
pydantic_core                            2.46.4
pydantic-extra-types                     2.11.2
pydantic-settings                        2.14.2
pyext                                    0.5
Pygments                                 2.20.0
pygtrie                                  2.5.0
PyJWT                                    2.13.0
pylatexenc                               2.11
pyparsing                                3.3.2
pyproject_hooks                          1.2.0
pypto                                    0.2.1
pytest                                   8.3.2
pytest-xdist                             3.6.1
python-dateutil                          2.9.0.post0
python-dotenv                            1.2.2
python-json-logger                       4.1.0
python-Levenshtein                       0.27.3
python-multipart                         0.0.32
PyYAML                                   6.0.1
pyzmq                                    27.1.0
Quart                                    0.20.0
rank-bm25                                0.2.2
RapidFuzz                                3.14.5
ray                                      2.48.0
referencing                              0.37.0
regex                                    2026.7.10
remote-pdb                               2.1.0
requests                                 2.34.2
retrying                                 1.4.2
rich                                     15.0.0
rich-toolkit                             0.20.3
rignore                                  0.8.0
rouge                                    1.0.1
rouge-chinese                            1.0.3
rouge-score                              0.1.2
rpds-py                                  2026.6.3
sacrebleu                                2.6.0
safetensors                              0.8.0
schedule_search                          0.0.1
scikit-image                             0.25.2
scikit-learn                             1.9.0
scipy                                    1.13.1
seaborn                                  0.13.2
sentencepiece                            0.2.2
sentry-sdk                               2.66.0
setproctitle                             1.3.7
setuptools                               84.0.0
setuptools-scm                           10.2.0
sgl_kernel_npu                           2026.6.1
sglang                                   0.0.0.dev15999+g91e7568c4.d20260804 /data2/sglang-0803/sglang/python
shellingham                              1.5.4
show_kernel_debug_data                   0.1.0
simplejson                               4.1.1
six                                      1.17.0
smg-grpc-proto                           0.4.14
smg-grpc-servicer                        0.8.0
sniffio                                  1.3.1
sortedcontainers                         2.4.0
soundfile                                0.13.1
sse-starlette                            3.4.5
stack-data                               0.6.3
starlette                                0.50.0
superkernel                              0.1.0
supervisor                               4.3.0
sympy                                    1.14.0
tabulate                                 0.10.0
te                                       0.4.0
termcolor                                3.3.0
threadpoolctl                            3.6.0
tifffile                                 2026.3.3
tiktoken                                 0.13.0
timeout-decorator                        0.5.0
timm                                     1.0.16
tokenizers                               0.22.2
torch                                    2.10.0+cpu
torch_npu                                2.10.0.post4.dev20260715
torchao                                  0.9.0+cpu
torchaudio                               2.10.0+cpu
torchvision                              0.25.0+cpu
tqdm                                     4.69.0
traitlets                                5.16.1
transformers                             5.12.1
tree-sitter                              0.21.3
tree-sitter-languages                    1.10.2
trimesh                                  5.0.0
triton                                   3.2.0
triton_ascend                            3.2.2
typer                                    0.27.0
typing_extensions                        4.16.0
typing-inspection                        0.4.2
urllib3                                  2.7.0
uvicorn                                  0.51.0
uvloop                                   0.22.1
vcs-versioning                           2.2.2
watchfiles                               1.2.0
wcmatch                                  11.0.1
wcwidth                                  0.8.2
websockets                               16.1.1
Werkzeug                                 3.1.8
wheel                                    0.47.0
word2number                              1.1
wsproto                                  1.3.2
xatlas                                   0.0.11
xgrammar                                 0.2.1
xxhash                                   3.8.1
yapf                                     0.43.0
yarl                                     1.24.2
zhconv                                   1.4.3
zipp                                     4.1.0
zope.event                               6.2
zope.interface                           8.5
